### PR TITLE
Patch fix store

### PR DESCRIPTION
**Pull request by jtclemm**
**Summary**

Fix store uses a different method to pack the size of peratom data in restart files than other fixes (such as property/atom, spring/self, neighbor/history, etc). From my understanding, this can cause one fix to read data associated with another fix, as seen in the example input scripts below. Replacing the method of casting integers in fix store resolves the problem.

Typically when a restart file is written, the size of peratom data is written directly to a buffer as an integer (e.g. buf[0] = nvalue+1). When the restart is read, the size of the peratom data (followed by the actual data) is temporarily stored in an array, atom->extra. Each fix with a registered restart callback will then locate the position of its data in atom->extra by incrementing a counter m using:
for (int i = 0; i < nth; i++) m += static_cast\<int\> (extra[nlocal][m]).
where nth is the number of fixes which wrote data prior this fix.

However, fix/store is an exception and saves a value of ubuf(nvalue+1).d to the buffer and casts the size of peratom data chunks back to an integer using (int) ubuf(extra[nlocal][m]).i. I don't fully understand how ubuf casts data (maybe I am incorrect), but testing seems to indicate the two methods are incompatible. If a simulation uses fix/store with another fix that stores peratom data, the fixes will fail to locate the correct position of their data in the extra array and will read incorrect data. 

**Related Issues**

None

**Author(s)**

Joel Clemmer - Sandia National Labs

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Old restart files that saved data from 2 or more instances of fix store will no longer be read correctly.

**Implementation Notes**

Updated syntax in fix store to match property/atom, spring/self, neigh/history, etc. Packing the size of data in restart files is now uniform.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

Attached are two minimal input files which reproduce the bug on my system.

An LJ melt simulation is initialized in input_1.txt. A dump file is written with two columns corresponding to the initial y position (stored in a store/state fix) and the initial x position (calculated from a displace/atom compute). Note that compute displace/atom uses fix/store.

The simulation is restarted in input_2.txt. After restarting, fix store and fix store/state fail to read the correct peratom data (both read the first data written) and the two columns in the dump correspond to the reference y position.

[input_1.txt](https://github.com/lammps/lammps/files/5038279/input_1.txt)
[input_2.txt](https://github.com/lammps/lammps/files/5038281/input_2.txt)

### CODE REVIEW

**Comment by akohlmey** [comment]
@jtclemm Please note that 
```
int ione = 1;
double dval = ubuf(ione).d;
int itwo = (int) ubuf(dval).i;
```
and
```
int ion = 1;
double dval = ione;
int itwo = (int) dval;
```
should be producing the same result in `itwo`. The difference is that the first method also works for all 64-bit integers (`int64_t` or `bigint`), while the second only if the value is less than 2^52, due to the lower precision of double precision numbers since the remaining bits are used for the exponent.

@sjplimp any idea why this makes a difference?

**Comment by jtclemm** [comment]
Hi Axel, thank you for the reply. So ubuf is used to avoid issues converting 32 and 64 bit datatypes?

My understanding is that the bug is equivalent to this case.
```
int ione = 1;
double dval = ubuf(ione).d;
int itwo = (int) dval;
```
I don't quite follow the casting, but testing I find itwo = 0. I'll look up ubuf and see if I can figure out what it's doing to confirm.

If only fix stores are used, data is restarted without errors. However, if fix store is used with another fix (say property/atom) data is then read incorrectly. I think this is because fix store saves/reads the data size using ubuf while all other fixes just use static_cast\<int\>. In my tests, this mismatch of ubuf returns zero.

**Comment by akohlmey** [comment]
> Hi Axel, thank you for the reply. So ubuf is used to avoid issues converting 32 and 64 bit datatypes?
> 
> My understanding is that the bug is equivalent to this case.
> 
> ```
> int ione = 1;
> double dval = ubuf(ione).d;
> int itwo = (int) dval;
> ```

there is a big difference between `(int) dval` and `(int) ubuf(dval).i`.

To explain. An IEEE754 double precision floating point number has:
- 52 bits mantissa
- 11 bits exponent
- 1 bit sign
Everything is base 2 and the bits in the mantissa represent 1/2 + 1/4 + 1/8 + 1/16 and so on. to that you always add 1, so the mantissa represents numbers between 1 and 2.
this is multiplied by 2^exponent. that way if the mantissa would become larger than 2, all bits are shifted and the exponent incremented by 1. and so on. there is a special case, if all bits are 0: that gives you 0.0 (which is different from -0.0, BTW, since that has the sign bit set). Because of this setup you can represent all integers up to 2^52 exactly but, for example, something like 0.1 not. However, when using the union, we can also use the bits for the exponent and the sign to store integers. thus `ubuf(1).d` becomes -0.0 as now the sign bit is set but all other bits are still 0.

or put differently with ubuf(int i), the value of i is promoted to an int64_t and then stored in the same storage as a double. with .d we copy the bitpattern (without looking at its value) into the communication buffer, which only stores doubles. on the reverse step, you initialize the union with the double bitpattern but with the .i you access it as an int64_t, which recovers the original value. the (int) cast is not really needed and primarily used to silence compiler warnings about truncation.

You can find the definition of ubuf and some (inline) docs in the file lmptype.h
 
> I don't quite follow the casting, but testing I find itwo = 0. I'll look up ubuf and see if I can figure out what it's doing to confirm.

what is your compiler?

What you are doing *should* not make a difference, but apparently it does. I don't see why.  If this would not work, LAMMPS would not be able to run. In fact, we recently had some hard to find bugs, where the ubuf(val).i was missing and just val was used.

> If only fix stores are used, data is restarted without errors. However, if fix store is used with another fix (say property/atom) data is then read incorrectly. I think this is because fix store saves/reads the data size using ubuf while all other fixes just use static_cast<int>. In my tests, this mismatch of ubuf returns zero.

**Comment by jtclemm** [comment]
Thank you for the great explanation, it was very helpful. I also looked at lmptype.h and now see the definition of ubuf, I didn't know it was a union. 

I'm still internalizing the details of ubuf, but to make sure I'm on the same page, what changes were you referring to that shouldn't make a difference? The change I made in the pull request, the change I made in the snippet of test code you provided (`int itwo = (int) ubuf(dval).i;` -> `int itwo = (int) dval;`), or both changes? 

I tested on two machines, one used g++ 4.8.5 and the other was g++ 7.5.0 (on Windows subsystem for Linux).

**Comment by akohlmey** [comment]
> I'm still internalizing the details of ubuf, but to make sure I'm on the same page, what changes were you referring to that shouldn't make a difference? The change I made in the pull request, the change I made in the snippet of test code you provided (`int itwo = (int) ubuf(dval).i;` -> `int itwo = (int) dval;`), or both changes?

the changes in your patch should not make a difference. Your change to my snippet is incorrect.

**Comment by jtclemm** [comment]
Okay that helps clarify. I better understand ubuf now and I agree that the deliberate change I made to to the snippet is incorrect.

I believe the problem with using fix store with another fix like property/atom is analogous to this snippet. In the (un)pack_restart methods in fix property atom, ubuf is NOT used to pack  or unpack the size of data chunks (lines 582 and 607). So when fix property atom tries to read the size of data associated with fix store, it directly converts what fix store wrote, `ubuf(nvalues+1).d`, to an integer and fails. Unless there is something else I'm missing. 

If integers should always be packed using ubuf, I can alternatively submit a pull request adding the appropriate ubuf calls to all the other fixes with restart atom callbacks. I believe that includes:
move, neigh/history, property/tom, spring/self, store/state, wall/gran, wall/gran/region, gld, ttm, peri_neighbor, gle, pimd, ti/spring, ttm/mod, meso/move
This should also fix the bug, leaving fix store the way it is.

Thanks again for the help!

**Comment by akohlmey** [comment]
@jtclemm it should be handled consistently. the main purpose of using the ubuf() packing is to support large systems where the number of atoms is > 2^31, but for that the corresponding "nvalues" variables also have to be defined as bigint.
However, this is touching a very sensitive part of the core of LAMMPS, so I won't make any suggestion unless we also have the input from @sjplimp, who has the final word on such changes and should be consulted before doing any changes of this kind.

Bottom line, inconsistency is bad. But if we revise it, we should consider making it compatible and consistent with the rest of LAMMPS as much as possible.

As I hinted at in a previous comment, these kind of omissions can happen and sometimes can go unnoticed for a while. We are incrementally adding a more thorough and low-level checking of LAMMPS internal procedures in the unittest tree, which has been helping enormously to detect corner cases or rarely used combinations of steps.

**Comment by sjplimp** [comment]
@akohlmey @jtclemm While I know that some classes pack ints with ubuf() and some don't, the part of this I'm not following is this:

I believe the problem with using fix store with another fix like property/atom is analogous to this snippet. In the (un)pack_restart methods in fix property atom, ubuf is NOT used to pack or unpack the size of data chunks (lines 582 and 607). So when fix property atom tries to read the size of data associated with fix store, it directly converts what fix store wrote, ubuf(nvalues+1).d, to an integer and fails. Unless there is something else I'm missing.

I'm not clear when it's ever the case that one fix would unpack something that another fix packed.
In other words, so long as each fix (or other class) does both it's pack and unpack consistently (with or w/out ubuf)
that there should be no problem.  What am I missing?

**Comment by akohlmey** [comment]
@sjplimp there are two issues here:
1.  why does the patch have an effect, even though it should not?
2.  what is the issue with the two inputs attached to the description at the top not producing consistent results and why does the patch "cure" it

**Comment by jtclemm** [comment]
Hi Steve. It took me a while to trace the evaluation order to track down the root cause of this bug and I may have missed some details. If I understand correctly, when a fix first packs its data in the restart buffer, it first writes how many values belong to it (say `nvalues+1`) then proceeds to write those values. When the restart file is read, all of this data for every fix is transferred into `atom->extra` which is individually parsed by each fix when it is reinitialized.

To reload its data,  a fix needs to identify the location of its data in `atom->extra`. It does this by first knowing its position in the list of atom restart call backs (`nth`) which tells it how many fixes wrote data before it. It then strides through `atom->extra` by reading the number of values written by the first fix, skipping that chunk of data, finding the size of data of the next fix, skipping that chunk, etc, until it reaches its own data. This is accomplished in this loop:
`for (int i = 0; i < nth; i++) m += static_cast<int> (extra[nlocal][m]).`

This can fail when a fix incorrectly parses the size of data written by another fix (`nvalues+1`) when one fix uses ubuf (only done in fix store) while they don't. Or vice versa.

**Comment by sjplimp** [comment]
ah, I see.  I forgot about that extra int that fixes have to scan over from all the other fixes when
they unpack restart data.  Good find!

So is the solution simply that all fixes need to pack/unpack that one integer consistently?
Not just consistent within the fix, but done the same way by all fixes?

And if fix store is the only one that is inconsistent, then it should be changed to 
buf[0] = nvalues+1;    // pack
for (int i = 0; i < nth; i++) m += static_cast<int> (extra[nlocal][m]);  // unpack

If that is what is in your patch, then I'm not sure what @akohlmey meant
by his last comment about why the patch has an effect ?

If we want to make sure the ubuf() logic is used everywhere for everything
(no static cast to ints out of bufs), that would also fix the issue, but that's a bigger job.

**Comment by sjplimp** [comment]
Undeed one of those indeeds ...

**Comment by akohlmey** [comment]
> If that is what is in your patch, then I'm not sure what @akohlmey meant
> by his last comment about why the patch has an effect ?

because I was assuming that packing unpacking of data is local to the fix only.
so the clean approach would be to also add a comment to all equivalent locations that the storing of the integer data must be done exactly this one way so that other fixes can read it.

> If we want to make sure the ubuf() logic is used everywhere for everything
> (no static cast to ints out of bufs), that would also fix the issue, but that's a bigger job.

i don't think that ubuf() would be needed here because "m" would overflow and all the pack/unpack routines use 32-bit ints, too.

axel.

**Comment by sjplimp** [comment]
> so the clean approach would be to also add a comment to all equivalent locations that the storing of the integer data must
> be done exactly this one way so that other fixes can read it.

Good idea.  That would hopefully prevent someone (probably me in the current instance) from making the
same mistake in a new fix in the future.

**Comment by jtclemm** [comment]
I'm happy to add such a comment to all the relevant fixes. I already have a list of them so it won't take me long. I can then just update this pull request. Maybe state something like:
// Note: the number of values packed to the buffer needs to be read and parsed by other fixes so it must be stored in this precise way.

**Comment by sjplimp** [comment]
Less is more.  I suggest:
// pack buf[0] this way b/c other fixes unpack it
// unpack the Nth first values this way b/c other fixes pack them

**Comment by jtclemm** [comment]
Hey everyone, I added the proposed comments to every fix with a (un)pack restart method. I also added `static_cast<int>` to USER-MISC/fix_srp.cpp pack_restart (line 578) to keep everything uniform. My understanding is this won't make a difference but I can also revert it back.

**Comment by akohlmey** [comment]
@jtclemm reverted the permission change and replaced 'b/c' with 'because'.
@sjplimp will merge this as soon as it passes the integration tests. any further tweaks or related changes can be in a new PR.
